### ISSUE TITLE
Add support for default arguments in UDFs

### ISSUE THREAD

**Issue by imor** [supabase/pg_graphql#410]
The scope of the [PR to add basic support for UDFs](https://github.com/supabase/pg_graphql/pull/393) has excluded more features in the interests for getting basic support out. One of the missing pieces is adding support for default arguments. This issue is to keep track of that sub-feature.

**Comment by olirice** [comment]
> Add support for default arguments

this is done, right?

> Add support for UDFs with unnamed arguments

(imo) this one can be skipped

> Add support for overloaded functions.

For those one we can probably require a comment directive `@graphql({"name": "foo"})` to disambiguate the names to get them to show up in the API

> Better support for default arguments in __schema

Can skip for now. as you mentioned, similar software made that choice

> Add support for functions returning void.

Can skip for now. It's a low lift to ask users to return a value e.g. `return 1;` so we can have a meaningful type in the schema (`null` is not a valid type so we'd otherwise have to pick one at random)

> Add support for functions accepting or returning array types.

+1

**Comment by ahoopes16** [comment]
Being able to support array arguments would be amazing. I am currently attempting to write a custom query so that I can filter records from my `courses` table based on a foreign relationship to our `authors` table. I need to be able to filter by multiple authors at once, so we were attempting to make our custom function with an `author_ids uuid[]` argument.

If there's anything that could be done to help, please let me know! Thanks so much, these custom functions are going to be incredibly useful!

**Comment by imor** [comment]
@ahoopes16 support for array types is one of the next items we plan to tackle for pg_graphql. For now you can simulate multiple filter arguments by e.g. using a single argument with comma separated values.

**Comment by ahoopes16** [comment]
Yep, that's exactly the workaround that we came to as well. Thanks!

**Comment by tylerdmace** [comment]
@imor @ahoopes16 I don't see any supported string splitting functions available to my functions. How are you taking a comma delimited string and splitting that to be used for a "where" clause? Can you show me an example of a simple function that does this?

**Comment by ahoopes16** [comment]
@tylerdmace Sure thing! So for this example, let's say that we have a `courses` table with a field called `author_id`. If you want to support a multi-select filter for courses by authors (as we did), you can write a UDF like this:

```sql
create or replace function search_courses(author_ids text default null)
returns setof courses
stable
language plpgsql
as
$$
declare
begin
    return query select c.* from courses c
    where
        search_courses.author_ids is null or
        c.author_id::text = any(string_to_array(search_courses.author_ids, ','));
end
$$;
```

The piece that you're looking for specifically is the `string_to_array(<comma-separated string parameter>, ',')`. I hope this helps!

**Comment by tylerdmace** [comment]
> I hope this helps!

This was exactly what I was looking for. I swear I had tried to use `string_to_array()` in the past and it wasn't recognized as an internal function that I could use and threw an error. But I'll revisit this tonight and see if I can get it working. Thank you!

**Comment by ahoopes16** [comment]
Strange! It's been working great for us on Postgres 15, although `string_to_array()` was introduced quite a few versions ago so I don't think version should matter here. Either way, glad I could help and I hope you get it working!

**Comment by tylerdmace** [comment]
It worked perfectly! Thank you so much, my friend.

**Comment by amacfie** [comment]
Thanks for the work so far and I've done a small one-time sponsorship to support further work on this issue.
